### PR TITLE
feat(call): multi speaker view

### PR DESCRIPTION
**Pull request by DorraJaouad**
## ☑️ Resolves

* Part of #18905

It is a big PR bc it is a giant feature. The logic behind multi speaker is to use automation based on speaking. The demotion happens after 1 min of silence. It is max 4 speakers. 

### AI (if applicable)

- [x] The content of this PR was partly or fully generated using AI

I prepared a simulation in the last commit to help test, you can alter it to test edge cases of course (I included the case of 5 to 6 speakers).

### 🖼️ Screenshots / Screencasts

https://github.com/user-attachments/assets/5709f239-0f1a-4dea-9632-c82871f1e829

### 🏁 Checklist

- [ ] 🌏 Tested with different browsers / clients:
  - [x] Chromium (Chrome / Edge / Opera / Brave)
  - [ ] Firefox
  - [ ] Safari
  - [ ] Talk Desktop
  - [ ] Integrations with Files sidebar and other apps
  - [ ] Not risky to browser differences / client
- [ ] 🖌️ Design was reviewed, approved or inspired by the design team
- [ ] ⛑️ Tests are included or not possible
- [ ] 📗 User documentation in https://github.com/nextcloud/documentation/tree/master/user_manual/talk has been updated or is not required

### CODE REVIEW

File: src/components/CallView/useActiveSpeakers.ts [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+ * SPDX-FileCopyrightText: 2026 Nextcloud GmbH and Nextcloud contributors
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+import type { Ref } from 'vue'
+
+import { computed, onScopeDispose, ref, watch } from 'vue'
+
+/** Time (in ms) a participant has to speak without interruption to be promoted. */
+export const PROMOTE_DELAY = 3000
```
Comment (Antreesy) [inline src/components/CallView/useActiveSpeakers.ts]:
```suggestion
export const PROMOTE_DELAY = 3_000
```

File: src/components/CallView/useActiveSpeakers.ts [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+/** Time (in ms) a promoted participant stays silent before being demoted again. */
+export const DEMOTE_DELAY = 60000
```
Comment (Antreesy) [inline src/components/CallView/useActiveSpeakers.ts]:
```suggestion
export const DEMOTE_DELAY = 60_000
```

File: src/components/CallView/useActiveSpeakers.ts [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+		if (!enabled.value) {
+			return []
+		}
+
+		return callParticipantModels.value
+			.filter((model) => model.attributes.speaking)
+			.map((model) => model.attributes.nextcloudSessionId)
+	})
+
+	watch(speakingSessionIds, (sessionIds, previousSessionIds) => {
```
Comment (Antreesy) [inline src/components/CallView/useActiveSpeakers.ts:140]: We kind of already collecting this (or close to this) information in `useTileOrdering.ts` composable and `participantActivity.ts` store, can it be reused?

File: src/components/CallView/useActiveSpeakers.spec.ts [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+
+			for (const model of models) {
+				model.attributes.speaking = true
+				await advance(PROMOTE_DELAY)
+			}
+
+			expect(promotedSessionIds.value).toEqual(['session-0', 'session-1', 'session-2', 'session-3'])
+		})
+
+		test('lets a queued speaker in once a slot frees up', async () => {
```
Comment (Antreesy) [inline src/components/CallView/useActiveSpeakers.spec.ts]: That's probably a highly unlikely edge case, when 5 people are speaking at the same time, and we need to queue fifth person.
But 5 people speaking within a minute is reasonable interval, in my opinion we should assume some of first four people no longer speaks.
Either we should:
- make it simple and kick out the oldest (double-check, but `demoteTimers[sessionId]` should store incremental numeric id of the timeout - the lower the value, the older the timer)
- slightly overkill the solution and add 10 seconds 'ease' period - if oldest haven't speak for 10+ seconds, kick them, otherwise queue the fifth person, then kick after ease period ended

File: src/components/CallView/Grid/SpeakersGrid.vue [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+const videoCount = computed(() => props.models.length)
+
+// This grid only ever holds promoted speakers, the local video lives in the
+// stripe below it.
+const noLocalVideoReserve = ref(true)
+
+const {
+	columns,
+	rows,
```
Comment (Antreesy) [inline src/components/CallView/Grid/SpeakersGrid.vue:45]: Shouldn't SpeakersGrid be locked to a single row? I'd expect it to be a stripe with large tiles, fitting next to each other 
Discussed with Nimisha, max 3, if possible, max 2 when mobile

This looks weird, basically nobody's promoted
Image

VideoVue also needs adjustments, 'Vanessa' tile is promoted, and video assumed to be on speakers area already

File: src/components/CallView/Grid/tilePromotionTransition.ts [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+ */
+function findTile(sessionId: string): HTMLElement | null {
+	return document.querySelector(`[${TILE_SESSION_ATTRIBUTE}="${CSS.escape(sessionId)}"]`)
+}
+
+/**
+ * Whether the user asked not to be shown motion.
+ */
+function prefersReducedMotion(): boolean {
+	return window.matchMedia('(prefers-reduced-motion: reduce)').matches
```
Comment (Antreesy) [inline src/components/CallView/Grid/tilePromotionTransition.ts]: Should be a one-time check, or cached result? It's called on every tile swap

File: src/components/CallView/Grid/tilePromotionTransition.ts [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+	}
+
+	tile.style.transition = 'none'
+	tile.style.transformOrigin = 'top left'
+	tile.style.transform = `translate(${from.left - to.left}px, ${from.top - to.top}px)`
+		+ ` scale(${from.width / to.width}, ${from.height / to.height})`
+
+	// Apply the offset before transitioning it away, otherwise the tile is
+	// simply painted at its new place
+	tile.offsetWidth
```
Comment (Antreesy) [inline src/components/CallView/Grid/tilePromotionTransition.ts:100]: ?

File: src/components/CallView/Grid/tilePromotionTransition.ts [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+ *
+ * @param sessionIds - session ids of the tiles of the main area
+ * @param previousSessionIds - the same, before the change
+ */
+export function animateTilePromotion(sessionIds: string[], previousSessionIds: string[]) {
+	if (prefersReducedMotion()) {
+		return
+	}
+
+	const inMainArea = new Set(sessionIds)
```
Comment (Antreesy) [inline src/components/CallView/Grid/tilePromotionTransition.ts]: MainArea is SpeakersGrid? Let's rename to `promoted*` for consistency?

File: src/components/CallView/CallView.vue [hunk 9]
```diff
@@ -338,6 +339,16 @@ export default {
 				&& this.promotedSpeakerModels.length > 1
 		},
 
+		// Session ids of the tiles actually shown in the main area, empty when a
+		// single participant is promoted there as usual
+		mainAreaSessionIds() {
```
Comment (Antreesy) [inline src/components/CallView/CallView.vue]: now use this for `stripeParticipantModels` computed

File: src/components/CallView/speakerSimulation.ts [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+ * One turn of speech: who speaks, when they start (in seconds from the
+ * beginning of the timeline) and for how long.
+ *
+ * The timeline is built to exercise every rule of the layout, and loops.
+ */
+const TIMELINE: { speaker: number, at: number, duration: number }[] = [
+	// A holds the floor: promoted after 3s, alone in the main area
+	{ speaker: 0, at: 0, duration: 20 },
+	// B asks a question: the main area splits in two
+	{ speaker: 1, at: 6, duration: 6 },
```
Comment (Antreesy) [inline src/components/CallView/speakerSimulation.ts:42]: B promoted straight to the SpeakersGrid, after 3 seconds they only starting to share 'the floor'
Image

File: src/components/CallView/CallView.vue [hunk 11]
```diff
@@ -116,7 +125,7 @@
 					:isRecording="isRecording"
 					:token="token"
 					:isOverlap="showFullPage"
-					:callParticipantModels="callParticipantModels"
+					:callParticipantModels="stripeParticipantModels"
```
Comment (Antreesy) [inline src/components/CallView/CallView.vue:130]: VideosGrid can now receive only one model (because the rest are filtered) and give false positive on `isLessThanTwoVideos` computed

File: src/components/CallView/Grid/tilePromotionTransition.ts [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+	tile.style.transition = 'none'
+	tile.style.transformOrigin = 'top left'
+	tile.style.transform = `translate(${from.left - to.left}px, ${from.top - to.top}px)`
+		+ ` scale(${from.width / to.width}, ${from.height / to.height})`
+
+	// Apply the offset before transitioning it away, otherwise the tile is
+	// simply painted at its new place
+	tile.offsetWidth
+
+	tile.addEventListener('transitionend', reset)
```
Comment (Antreesy) [inline src/components/CallView/Grid/tilePromotionTransition.ts:102]: Other animations could also trigger 'transitionend' here and intercept?
`reset` probably need a check that it's a 'transform' animation that was ended; also maybe set 'pointerEvents' to 'none', so mouseover events do not trigger anything?

File: src/components/CallView/BottomBar.vue [hunk 13]
```diff
@@ -159,15 +160,44 @@ const fullscreenLabel = computed(() => {
...
+	return callViewStore.isMultiSpeaker ? CallLayout.MultiSpeaker : CallLayout.Speaker
+})
+
+const callLayoutLabel = computed(() => {
+	return callLayouts.value.find((layout) => layout.id === callLayout.value)!.label
+})
+
+const callLayoutIcon = computed(() => {
+	return callLayouts.value.find((layout) => layout.id === callLayout.value)!.icon
+})
```
Comment (Antreesy) [inline src/components/CallView/BottomBar.vue]: Was confused with enum for a moment. Let's simplify it

```suggestion
const CALL_LAYOUT = {
	GRID: 'grid',
	SPEAKER: 'speaker',
	MULTI_SPEAKER: 'multi-speaker',
} as const

const CALL_LAYOUT_COMPONENT = {
	[CALL_LAYOUT.GRID]: {
		label: t('spreed', 'Grid view'),
		icon: IconViewGridOutline,
	},
	[CALL_LAYOUT.SPEAKER]: {
		label: t('spreed', 'Speaker view'),
		icon: IconViewGalleryOutline,
	},
	[CALL_LAYOUT.MULTI_SPEAKER]: {
		label: t('spreed', 'Multi-speaker view'),
		icon: IconViewDashboardOutline,
	},
} as const

const callLayouts = [CALL_LAYOUT.GRID, CALL_LAYOUT.SPEAKER, CALL_LAYOUT.MULTI_SPEAKER] as const

const callLayout = computed(() => {
	if (isGrid.value) {
		return CALL_LAYOUT.GRID
	}
	return callViewStore.isMultiSpeaker ? CALL_LAYOUT.MULTI_SPEAKER : CALL_LAYOUT.SPEAKER
})
```

File: src/components/CallView/BottomBar.vue [hunk 14]
```diff
@@ -447,17 +486,28 @@ function changeView() {
...
 				variant="tertiary"
-				:aria-label="changeViewLabel"
-				:title="changeViewLabel"
-				@click="changeView">
+				:aria-label="callLayoutLabel"
+				:title="callLayoutLabel">
 				<template #icon>
-					<IconViewGridOutline v-if="!isGrid" :size="20" />
-					<IconViewGalleryOutline v-else :size="20" />
+					<component :is="callLayoutIcon" :size="20" />
```
Comment (Antreesy) [inline src/components/CallView/BottomBar.vue]: No need for computed in template
```suggestion
					<component :is="CALL_LAYOUT_COMPONENT[callLayout].icon" :size="20" />
```

File: src/components/CallView/CallView.vue [hunk 15]
```diff
@@ -285,8 +320,56 @@ export default {
...
 		},
 
-		callParticipantModels() {
-			return callParticipantCollection.callParticipantModels.filter((callParticipantModel) => !callParticipantModel.attributes.internal || callParticipantModel.attributes.videoAvailable)
+		// The speakers currently sharing the main area, in the order they became
+		// active
+		promotedSpeakerModels() {
+			return this.promotedSessionIds
+				.map((sessionId) => this.callParticipantModels.find((model) => model.attributes.nextcloudSessionId === sessionId))
+				.filter((model) => model !== undefined)
```
Comment (Antreesy) [inline src/components/CallView/CallView.vue:332]: Can this happen? Shouldn't already be filtered out by composable?

File: src/components/CallView/useActiveSpeakers.ts (hunk 3 above)
Comment (DorraJaouad) [reply src/components/CallView/useActiveSpeakers.ts:140]: It would require refactoring grid promotion path too, so I won't make this PR any larger and I will put it in a follow up

File: src/components/CallView/Grid/tilePromotionTransition.ts (hunk 7 above)
Comment (DorraJaouad) [reply src/components/CallView/Grid/tilePromotionTransition.ts:100]: Well, from my research, offsetWidth is defined in terms of the element's layout box, so its getter cannot answer without up-to-date layout. Reading it forces the browser to flush pending style changes and run layout synchronously

File: src/components/CallView/speakerSimulation.ts (hunk 10 above)
Comment (DorraJaouad) [reply src/components/CallView/speakerSimulation.ts:42]: There is an intentional delay of 3 sec, because some participant can just make a small sound by accident and hence shall not be considered speakers.

File: src/stores/callView.ts [hunk 16]
```diff
@@ -94,7 +97,9 @@ export const useCallViewStore = defineStore('callView', {
 				// not defined yet, default to grid view for group/public calls, otherwise speaker view
 				? [CONVERSATION.TYPE.GROUP, CONVERSATION.TYPE.PUBLIC].includes(conversation.type)
 				: gridPreference === 'true'
-			this.setCallViewMode({ token: conversation.token, isGrid, isStripeOpen: true })
+			// Speaker view shows a single speaker unless multi-speaker was picked before
+			const isMultiSpeaker = BrowserStorage.getItem(`callprefs-${conversation.token}-multispeaker`) === 'true'
+			this.setCallViewMode({ token: conversation.token, isGrid, isMultiSpeaker, isStripeOpen: true })
```
Comment (Antreesy) [inline src/stores/callView.ts:102]: Do we want `isMultiSpeaker` to be the new default?

File: src/components/CallView/CallView.vue [hunk 17]
```diff
@@ -538,9 +675,15 @@ export default {
 		callParticipantCollection.on('remove', this._lowerHandWhenParticipantLeaves)
 
 		subscribe('switch-screen-to-id', this._switchScreenToId)
+
+		// TODO: development only, remove before opening a pull request
+		setUpSpeakerSimulation(this.token)
```
Comment (Antreesy) [inline src/components/CallView/CallView.vue]: It was quite handy for debugging, esp. if we'll add a playwright tests later. Let's keep it but hide behind `if (OC.debug) {}`?

File: src/components/CallView/CallView.vue [hunk 18]
```diff
@@ -463,6 +584,22 @@ export default {
...
+		},
+
+		isMultiSpeaker() {
+			this._setPromotedParticipant()
+		},
+
+		promotedAreaSessionIds(sessionIds, previousSessionIds) {
+			// Runs before the grids are re-rendered, while the tiles are still
+			// where they are about to move from
+			animateTilePromotion(sessionIds, previousSessionIds)
```
Comment (Antreesy) [inline src/components/CallView/CallView.vue:599]: Note: animations do work on Single-Speaker mode. But I dont' mind, it looks nice there

File: src/components/CallView/useActiveSpeakers.ts [hunk 19]
```diff
@@ -0,0 +1,209 @@
...
+	 */
+	function activate(sessionId: string, speaking: Set<string>) {
+		if (activeSessionIds.value.includes(sessionId)) {
+			return
+		}
+
+		if (activeSessionIds.value.length >= MAX_PROMOTED_SPEAKERS) {
+			const now = Date.now()
+			const recencyOf = (id: string) => (speaking.has(id) ? now : (lastSpokeAt[id] ?? 0))
+			const leastRecent = activeSessionIds.value
```
Comment (Antreesy) [inline src/components/CallView/useActiveSpeakers.ts:109]: Maybe return after these, if all 3 are still speaking? so No4 doesn't get promoted by interrupting No1-3. Simpe return without queuing

File: src/components/CallView/Grid/useGridDimensions.ts [hunk 20]
```diff
@@ -82,10 +86,6 @@ export function useGridDimensions({
 	const targetAspectRatio = computed(() => getTargetAspectRatio(isStripe.value))
 	const gridAspectRatio = computed(() => (gridWidth.value / gridHeight.value).toPrecision(2))
 
-	// The full grid reserves one slot for the local video, unless it is not shown
-	// (stripe or recording mode).
-	const noLocalVideoReserve = computed(() => isStripe.value || isRecording.value)
```
Comment (Antreesy) [inline src/components/CallView/Grid/useGridDimensions.ts:87]: We have a watcher `watch([isStripe, stripeOpen, isRecording], () => nextTick(recompute))` below that was tracking it indirectly. Let's replace `isRecording` with `noLocalVideoReserve`

File: src/components/CallView/CallView.vue [hunk 21]
```diff
@@ -281,12 +318,96 @@ export default {
...
+			}
+
+			return this.promotedParticipantModel
+				? [this.promotedParticipantModel.attributes.nextcloudSessionId]
+				: []
+		},
+
+		// Peer ids of the speakers actually shown in the main area, which deserve
+		// the same video quality as a single promoted participant
+		promotedSpeakerPeerIds() {
```
Comment (Antreesy) [inline src/components/CallView/CallView.vue:392]: Since we won't have more than ~3 promoted speakers, Set is overkill and it's fine to simplify this to Array and Array.includes()

**Review by Antreesy — APPROVED** [review]
Overall fine, left last comments

File: src/components/CallView/useActiveSpeakers.ts (hunk 19 above)
Comment (DorraJaouad) [reply src/components/CallView/useActiveSpeakers.ts:109]: Right now, number 4 replaces immediately the least recent speaker and not gets queued

File: src/components/CallView/CallView.vue (hunk 21 above)
Comment (DorraJaouad) [reply src/components/CallView/CallView.vue:392]: it is not effort-heavy overkill

File: src/components/CallView/useActiveSpeakers.ts (hunk 19 above)
Comment (Antreesy) [reply src/components/CallView/useActiveSpeakers.ts:109]: yes, what i wrote =P
but ok either way
